File: src/test/java/org/trp/shincolle/item/AbyssNuggetItemTest.java
package org.trp.shincolle.item;

import net.minecraft.network.chat.contents.TranslatableContents;
import net.minecraft.world.item.CreativeModeTab;
import net.minecraft.world.item.ItemStack;
import org.trp.shincolle.init.ModItems;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

class AbyssNuggetItemTest {

    @Test
    void abyssNuggetShouldClampVariantsAndExposeBothCreativeStacks() {
        AbyssNuggetItem item = (AbyssNuggetItem) ModItems.ABYSS_NUGGET.get();

        ItemStack base = item.createVariantStack(-1);
        ItemStack variant = item.createVariantStack(1);
        ItemStack clamped = item.createVariantStack(99);

        assertEquals(0, item.getVariant(base));
        assertEquals(1, item.getVariant(variant));
        assertEquals(1, item.getVariant(clamped));
        assertEquals(0, item.getModelVariant(base));
        assertEquals(1, item.getModelVariant(variant));

        List<ItemStack> stacks = new ArrayList<>();
        item.addAllVariantsToCreativeTab(new CollectingOutput(stacks));
        assertEquals(2, stacks.size());
        assertEquals(0, item.getVariant(stacks.get(0)));
        assertEquals(1, item.getVariant(stacks.get(1)));
    }

    @Test
    void abyssNuggetShouldUseLegacyTranslationKeysPerVariant() {
        AbyssNuggetItem item = (AbyssNuggetItem) ModItems.ABYSS_NUGGET.get();

        String baseKey = ((TranslatableContents) item.getName(item.createVariantStack(0)).getContents()).getKey();
        String variantKey = ((TranslatableContents) item.getName(item.createVariantStack(1)).getContents()).getKey();

        assertEquals("item.shincolle.AbyssNugget.name", baseKey);
        assertEquals("item.shincolle.AbyssNugget1.name", variantKey);
    }

    private record CollectingOutput(List<ItemStack> stacks) implements CreativeModeTab.Output {
        @Override
        public void accept(ItemStack stack, CreativeModeTab.TabVisibility visibility) {
            stacks.add(stack.copy());
        }
    }
}

File: src/test/kotlin/org/trp/shincolle/item/AbyssNuggetItemTest.kt
package org.trp.shincolle.item

import net.minecraft.network.chat.contents.TranslatableContents
import net.minecraft.world.item.CreativeModeTab
import net.minecraft.world.item.ItemStack
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.trp.shincolle.init.ModItems

class AbyssNuggetItemTest {

    @Test
    fun abyssNuggetShouldClampVariantsAndExposeBothCreativeStacks() {
        val item = ModItems.ABYSS_NUGGET.get() as AbyssNuggetItem

        val base = item.createVariantStack(-1)
        val variant = item.createVariantStack(1)
        val clamped = item.createVariantStack(99)

        assertEquals(0, item.getVariant(base))
        assertEquals(1, item.getVariant(variant))
        assertEquals(1, item.getVariant(clamped))
        assertEquals(0, item.getModelVariant(base))
        assertEquals(1, item.getModelVariant(variant))

        val stacks = mutableListOf<ItemStack>()
        item.addAllVariantsToCreativeTab(CollectingOutput(stacks))
        assertEquals(2, stacks.size)
        assertEquals(0, item.getVariant(stacks[0]))
        assertEquals(1, item.getVariant(stacks[1]))
    }

    @Test
    fun abyssNuggetShouldUseLegacyTranslationKeysPerVariant() {
        val item = ModItems.ABYSS_NUGGET.get() as AbyssNuggetItem

        val baseKey = (item.getName(item.createVariantStack(0)).contents as TranslatableContents).key
        val variantKey = (item.getName(item.createVariantStack(1)).contents as TranslatableContents).key

        assertEquals("item.shincolle.AbyssNugget.name", baseKey)
        assertEquals("item.shincolle.AbyssNugget1.name", variantKey)
    }

    private class CollectingOutput(private val stacks: MutableList<ItemStack>) : CreativeModeTab.Output {
        override fun accept(stack: ItemStack, visibility: CreativeModeTab.TabVisibility) {
            stacks += stack.copy()
        }
    }
}
